src/index.ts

```diff
@@ -1,65 +1,127 @@
+			if (!upgradeHeader || upgradeHeader !== 'websocket') {
+				return new Response('Durable Object expected Upgrade: websocket', {
+					status: 426,
+				});
+			}
+
+			if (request.method !== 'GET') {
+				return new Response('Durable Object expected GET method', {
+					status: 400,
+				});
+			}
+
+			// This example will refer to the same Durable Object,
+			// since the name "foo" is hardcoded.
+			let id = env.WEBSOCKET_HIBERNATION_SERVER.idFromName('foo');
+			let stub = env.WEBSOCKET_HIBERNATION_SERVER.get(id);
+
+			return stub.fetch(request);
+		}
+
+		return new Response(null, {
+			status: 400,
+			statusText: 'Bad Request',
+			headers: {
+				'Content-Type': 'text/plain',
+			},
+		});
+	},
+};
+
+// Durable Object
 export class WebSocketHibernationServer extends DurableObject {
-	/**
-	 * The constructor is invoked once upon creation of the Durable Object, i.e. the first call to
-	 * 	`DurableObjectStub::get` for a given identifier (no-op constructors can be omitted)
-	 *
-	 * @param ctx - The interface for interacting with Durable Object state
-	 * @param env - The interface to reference bindings declared in wrangler.toml
-	 */
+	// Keep track of all WebSocket connections
+	sessions: Map<WebSocket, any>;
+
 	constructor(ctx: DurableObjectState, env: Env) {
 		super(ctx, env);
+		this.sessions = new Map();
+
+		// DO saves the connection information when it goes to hibernation
+		// Get all WebSocket connections from the DO
+		this.ctx.getWebSockets().forEach((ws) => {
+			// Checks if this is a new DO
+			if (ws.deserializeAttachment()) {
+				// If a DO is waking up, deserialize the attachment.
+				// This helps to restore the state of the connection
+				const { ...session } = ws.deserializeAttachment();
+				this.sessions.set(ws, { ...session });
+			}
+		});
+
+		// Sets an application level auto response that does not wake hibernated WebSockets.
+		this.ctx.setWebSocketAutoResponse(new WebSocketRequestResponsePair('ping', 'pong'));
 	}
 
-	/**
-	 * The Durable Object exposes an RPC method sayHello which will be invoked when when a Durable
-	 *  Object instance receives a request from a Worker via the same method invocation on the stub
-	 *
-	 * @param name - The name provided to a Durable Object instance from a Worker
-	 * @returns The greeting to be sent back to the Worker
-	 */
-	async sayHello(name: string): Promise<string> {
-		return `Hello, ${name}!`;
+	async fetch(request: Request): Promise<Response> {
+		// Creates two ends of a WebSocket connection.
+		const webSocketPair = new WebSocketPair();
+		const [client, server] = Object.values(webSocketPair);
+
+		// Calling `acceptWebSocket()` informs the runtime that this WebSocket is to begin terminating
+		// request within the Durable Object. It has the effect of "accepting" the connection,
+		// and allowing the WebSocket to send and receive messages.
+		// Unlike `ws.accept()`, `state.acceptWebSocket(ws)` informs the Workers Runtime that the WebSocket
```

Comment:
`state.acceptWebSocket` doesn't make sense when we then write `this.ctx.acceptWebSocket()` below :wink:.